monitoring/src/monitoring_service/blockchain.py

```diff
@@ -0,0 +1,215 @@
+        web3: A Web3 instance
+        contract_manager: A contract manager
+        contract_address: The address of the contract to be filtered, can be `None`
+        contract_name: The name of the contract
+        topics: The topics to filter for
+        from_block: The block to start search events
+        to_block: The block to stop searching for events
+
+    Returns:
+        All matching events
+    """
+    filter_params = {
+        'fromBlock': from_block,
+        'toBlock': to_block,
+        'address': to_checksum_address(contract_address),
+        'topics': topics,
+    }
+
+    events = web3.eth.getLogs(filter_params)
+
+    return [
+        decode_event(
+            contract_manager.get_contract_abi(contract_name),
+            raw_event,
+        )
+        for raw_event in events
+    ]
+
+
+class BlockchainListener:
+    """ This is pull-based blockchain listener. """
+
+    def __init__(self, web3: Web3, contract_manager: ContractManager):
+        self.w3 = web3
+        self.contract_manager = contract_manager
+
+    def _get_token_network_registry_events(
+            self,
+            registry_address: str,
+            from_block: int,
+            to_block: int,
+    ) -> List[Dict]:
+        return query_blockchain_events(
+            web3=self.w3,
+            contract_manager=self.contract_manager,
+            contract_address=registry_address,
+            contract_name=CONTRACT_TOKEN_NETWORK_REGISTRY,
+            topics=create_registry_event_topics(self.contract_manager),
+            from_block=from_block,
+            to_block=to_block,
+        )
+
+    def _get_token_networks_events(
+            self,
+            network_address: str,
+            from_block: int,
+            to_block: int,
+    ) -> List[Dict]:
+        return query_blockchain_events(
+            web3=self.w3,
+            contract_manager=self.contract_manager,
+            contract_address=network_address,
+            contract_name=CONTRACT_TOKEN_NETWORK,
+            topics=create_channel_event_topics(),
+            from_block=from_block,
+            to_block=to_block,
+        )
+
+    def get_events(self, context: Context, to_block: int) -> Generator[Event, None, None]:
+        from_block = context.ms_state.latest_known_block
+
+        if to_block <= from_block:
+            return
+
+        log.info('Querying new block(s)', from_block=from_block, end_block=to_block)
+
+        # first check for new token networks
+        registry_events = self._get_token_network_registry_events(
+            registry_address=context.ms_state.token_network_registry_address,
+            from_block=from_block,
+            to_block=to_block,
+        )
+
+        for event in registry_events:
+            yield ReceiveTokenNetworkCreatedEvent(
+                token_network_address=event['args']['token_network_address'],
+            )
+
+        # then check all token networks
+        # the new token networks are registered by now
```

Comment:
This assumes that the `ReceiveTokenNetworkCreatedEvent`s are processed immediately on yield and added to context.ms_state, right? It's hard to get a proper overview in such a huge chage, but this might violate the abstraction between BlockchainListener and Monitoring service.

Reply:
Yes, this one is hard to decide. Right now it's kind of it's own event queue, that handles the whole state lifecycle and triggers events for all monitoring related tasks.

We can combine these, but then in the BC listener it becomes a bit more difficult to properly handle all events.

Reply:
I pushed some changes we discussed, can you have another look?